### PR TITLE
Add support for scikit-learn, XGBoost, lightGBM and CatBoost models

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request introduces classifiers for scikit-learn (LogisticRegression, SVC, LinearSVC, DecisionTreeClassifier, ExtraTreeClassifier, AdaBoostClassifier, BaggingClassifier, ExtraTreesClassifier,
GradientBoostingClassifier, RandomForestClassifier), XGBoost, lightGBM and CatBoost models.

It enables white-box attacks on `sklearn.linear_model.LogisticRegression` and `sklearn.svm.{SVC, LinerSVC}`, `sklearn.tree.DecisionTreeClassifier`, `GPy` (both @KaGrosse) and black-box attacks on all `sklearn` and `XGBoost`, `lightGBM` and `CatBoost` models.

It introduces a new white-box attack specific for DecisionTree classifiers (@KaGrosse)

It provides an example notebook for each newly supported classifier.

It introduces a new flexible API for all classifiers using abstract base classes for simple classifiers (black-box), neural networks, and gradient-providing classifiers (white-box).

Contributes to epic #47 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/boundary.py [hunk 1]
```diff
@@ -69,6 +70,12 @@ def __init__(self, classifier, targeted=True, delta=0.01, epsilon=0.01, step_ada
         :type init_size: `int`
         """
         super(BoundaryAttack, self).__init__(classifier=classifier)
+        if not isinstance(classifier, ClassifierNeuralNetwork) or not isinstance(classifier, ClassifierGradients):
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/boundary.py]: Yes, `BoundaryAttack` only requires `predict` and should only require `ClassifierGradients`. Currently `BoundaryAttack` seems to be implicitly implemented for image data in the lines following https://github.com/IBM/adversarial-robustness-toolbox/blob/dce8d74de74bc6560c0a0fbac3ee1d5d451431c6/art/attacks/boundary.py#L279 

I have a fix to make it work for 2-dimensional data, but I would like to dicuss it first with @minhitbk who implemented these lines.

File: art/classifiers/__init__.py [hunk 2]
```diff
@@ -2,9 +2,21 @@
...
 preexisting model.
 """
-from art.classifiers.classifier import Classifier
+from art.classifiers.classifier import Classifier, ClassifierNeuralNetwork, ClassifierGradients
 from art.classifiers.keras import KerasClassifier
 from art.classifiers.mxnet import MXClassifier
 from art.classifiers.pytorch import PyTorchClassifier
 from art.classifiers.tensorflow import TFClassifier
 from art.classifiers.ensemble import EnsembleClassifier
+from art.classifiers.scikitklearn import ScikitlearnLogisticRegression
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/__init__.py]: I think this is a great idea, I'm looking forward to the PR.

File: art/classifiers/scikitklearn.py [hunk 3]
```diff
@@ -0,0 +1,830 @@
...
+        for i_sample in range(num_samples):
+            for i_class in range(self.nb_classes):
+                gradients[i_sample, :] += class_weight[i_class] * (1.0 - y_preprocessed[i_sample, i_class]) * (
+                        self.weights[i_class, :] - w_weighted[i_sample, :])
+
+        gradients = self._apply_preprocessing_gradient(x, gradients)
+
+        return gradients
+
+    def predict(self, x, **kwargs):
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/scikitklearn.py]: I agree, it got forgotten when the base class was introduced. I removed it with 1e0763cdde72ab85acb3a062dec02cb81e9bd1be

File: art/classifiers/scikitklearn.py [hunk 4]
```diff
@@ -0,0 +1,830 @@
...
+
+        :param x: Test set.
+        :type x: `np.ndarray`
+        :return: Array of predictions of shape `(nb_inputs, self.nb_classes)`.
+        :rtype: `np.ndarray`
+        """
+        # Apply defences
+        x_preprocessed, _ = self._apply_preprocessing(x, y=None, fit=False)
+
+        return self.model.predict_proba(x_preprocessed)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/scikitklearn.py]: Yes, the majority of scikit-learn classifier models supported so far has `predict_proba`. For the case of `SVC` fitted without probability estimates or `LinearSVC` we overwrite `predict` in the sub-calss for SVC which checks if probability is available, and if not falls back to using the model's `predict`.

File: art/classifiers/scikitklearn.py [hunk 5]
```diff
@@ -0,0 +1,830 @@
...
+        if not isinstance(model, ExtraTreeClassifier):
+            raise TypeError(
+                'Model must be of type sklearn.tree.ExtraTreeClassifier')
+
+        super(ScikitlearnExtraTreeClassifier, self).__init__(model=model, clip_values=clip_values,
+                                                             defences=defences,
+                                                             preprocessing=preprocessing)
+
+
+class ScikitlearnAdaBoostClassifier(ScikitlearnClassifier):
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/scikitklearn.py]: It's correct that these classifiers currently only have 'fit' and 'predict' provided by the base class. For the tree-based classifiers we are already working on attacks and robustness metrics that will soon add more tree-specific methods to the tree-based classifiers, similar to `DecisionTreeClassifier`, which might eventually even lead to a new classifier API for tree-based classifiers. I think something similar will happen for generic ensemble models in the near future. Therefore I'd like to keep these classes as development space for future contributions.

I think connecting the generic ensemble models `AdaBoostClassifier` and `BaggingClassifier` with the ART classifiers for their specific base estimators is a great idea which should allow us to reuse a lot of code.

File: art/classifiers/scikitklearn.py [hunk 6]
```diff
@@ -0,0 +1,830 @@
...
+        # Apply preprocessing
+        x_preprocessed, y_preprocessed = self._apply_preprocessing(
+            x, y, fit=False)
+
+        num_samples, _ = x_preprocessed.shape
+        gradients = np.zeros(x_preprocessed.shape)
+
+        y_index = np.argmax(y_preprocessed, axis=1)
+        if self.model_class_weight is None or np.unique(y_index).shape[0] < self.nb_classes:
+            class_weight = np.ones(self.nb_classes)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/scikitklearn.py]: I am not completely sure if this is the perfect solution. It looks like class weights are only stored numerically in the model if they are provided as a dictionary to `LogisticRegression`. For the `None` option the `class_weights` are 1 and for the `balanced` option (currently covered by `np.unique(y_index).shape[0] < self.nb_classes` in line 577) they seem to be calculated on the fly during fitting. In `loss_gradient` we only have access to a sample of the labels and not the entire training data, therefore I think we cannot calculate class weights for the `balanced` option, therefore they are currently set to 1. 
I'm considering two option:
- removing `class_weight` entirely form `loss_gradient`
- or replace in line 577 `np.unique(y_index).shape[0] < self.nb_classes` with `self.model_class_weight == 'balanced'`

File: art/attacks/boundary.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (imolloy) [reply art/attacks/boundary.py]: Ah, ok. We could try to resolve it with this PR since it introduced the split in interfaces, or open an issue. Having it resolved together might be preferable.

File: art/classifiers/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (imolloy) [reply art/classifiers/__init__.py]: See #131

File: art/classifiers/scikitklearn.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (imolloy) [reply art/classifiers/scikitklearn.py]: Yes, I saw how it worked for SVC. Looking at sklearn in total though, I was actually surprised by what I saw: 

```python
from sklearn.utils.testing import all_estimators

estimators = all_estimators()

print("Has predict_proba")
for name, class_ in estimators:
    if hasattr(class_, 'predict_proba'):
        print(name)


print("\n\nMissing predict_proba")
for name, class_ in estimators:
    if not hasattr(class_, 'predict_proba'):
        print(name)
```

File: art/classifiers/scikitklearn.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (imolloy) [reply art/classifiers/scikitklearn.py]: Maybe the second option makes sense.

File: art/classifiers/scikitklearn.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/scikitklearn.py]: I think see your point. I have extended `ScikitlearnClassifier` with ebbcfab413f3edfbad00ef84232f6e5e0edfb236 accordingly, that way everything is accessible to black-box attacks. The second list "Missing predict_proba" also contains a lot of things that don't have `predict` either.

File: art/classifiers/scikitklearn.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/scikitklearn.py]: Resolved with cb0212abb2c00c5e1629854c0fba73ef2a2acdf4

File: art/classifiers/classifier.py [hunk 7]
```diff
@@ -57,35 +56,186 @@ def __init__(self, channel_index, clip_values=None, defences=None, preprocessing
...
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        attributes = {(k[1:], v) if k[0] == '_' else (k, v) for (k, v) in self.__dict__.items()}
+        attributes = ['{}={}'.format(k, v) for (k, v) in attributes]
+        repr_string = class_name + '(' + ', '.join(attributes) + ')'
+        return repr_string
+
+
+class ClassifierNeuralNetwork(ABC):
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/classifier.py]: I would vote to keep the three interfaces, because I can already recognise the need for fourth interface for tree-based classifiers in the near future. The `ClassifierNeuralNetwork` interface contains all the neural network specific functionalities like e.g. `learning_phase` or `get_activations` that are not needed in the any other interface. Only `Classifier` is a complete interface, whereas the other two interfaces can be mixed in with `Classifier`. That way `AdversarialPatch` requires both `Classifier` and `ClassifierNeuralNetwork` because it only makes sense for convolutional neural networks, as far as we know.

File: art/attacks/boundary.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/boundary.py]: Ok, I'll do some more testing and update BoundaryAttack soon.

File: art/classifiers/scikitklearn.py [hunk 8]
```diff
@@ -0,0 +1,830 @@
...
+        else:
+            self._input_shape = None
+
+    def fit(self, x, y, **kwargs):
+        """
+        Fit the classifier on the training set `(x, y)`.
+
+        :param x: Training data.
+        :type x: `np.ndarray`
+        :param y: Labels, one-vs-rest encoding.
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/scikitklearn.py]: I haven't checked all functions, but I think the vast majority expects one-vs-rest encoding as input and also provides predictions in this format. The `_apply_preprocessing` together with its gradient counterpart `_apply_preprocessing_gradient` have been recently introduced to summarise all preprocessing steps in the classifier interface to make sure they are applied in the same order across all classifiers.

File: art/classifiers/scikitklearn.py [hunk 9]
```diff
@@ -0,0 +1,830 @@
...
+
+        super(ScikitlearnLogisticRegression, self).__init__(clip_values=clip_values, defences=defences,
+                                                            preprocessing=preprocessing)
+
+        self.model = model
+        if hasattr(self.model, 'coef_'):
+            self.weights = self.model.coef_
+            self.classes = self.model.classes_
+            self._nb_classes = self.model.classes_.shape[0]
+            self.model_class_weight = self.model.class_weight
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/scikitklearn.py]: Very interesting thought. I made a test and I agree it's better to directly grab the weights from the model. Resolved with a14975c33d4209381c6e1787c0a7e365be84fa97

File: art/attacks/boundary.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/boundary.py]: Resolved in df504d916235e97ed68d204afc7cc435bce2168a

File: art/classifiers/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/classifiers/__init__.py]: I have merged #131 . Thank you.

File: art/attacks/boundary.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (imolloy) [reply art/attacks/boundary.py]: Isn’t it still assuming the model is an instance of `ClassifierGradients` now?

[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/boundary.py]: You are right, I didn't notice it because I tested it with `LogisticRegression` and `SVC`, which both have gradients. Resolved with 6231b2b6fe08c4373d877183827b1e52d5a857a4
